llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/IR/LinalgExtOps.cpp

```diff
@@ -1265,6 +1265,126 @@ LogicalResult TopkOp::verify() {
+  Value initialIndex = b.create<memref::LoadOp>(loc, indices(), ivs);
+
+  // Compute K (ub) from the selected dim of the output
+  Value ub;
+  ShapedType outputValuesType = outputValues().getType().cast<ShapedType>();
+  if (outputValuesType.isDynamicDim(dimension())) {
+    ub = b.create<memref::DimOp>(loc, outputValues(), dimension());
+  } else {
+    ub = b.create<arith::ConstantIndexOp>(
+        loc, outputValuesType.getDimSize(dimension()));
+  }
+
+  // Inner K loop functions:
+  //   Load current K val/ind
+  //   Compare N to K using inserted block compare
+  //   Check if N/K are equal, which came first
+  //   Store results in appropriate places
+  //   Yield loop carry vals
+  Value kValue, kIndex;
+  auto scfFor = b.create<scf::ForOp>(
+      loc, zero, ub, one, ValueRange{initialValue, initialIndex},
+      [&](OpBuilder &b, Location loc, Value iv, ValueRange loopCarryValues) {
+        SmallVector<Value> indices(ivs);
+        indices[kDim] = iv;
+        kValue = b.create<memref::LoadOp>(loc, outputValues(), indices);
+        kIndex = b.create<memref::LoadOp>(loc, outputIndices(), indices);
+      });
+
+  SmallVector<Value> indices(ivs);
+  indices[kDim] = scfFor.getInductionVar();
+  auto loopCarryValues = scfFor.getRegionIterArgs();
+
+  // Retrieve region as black box comparision function f(x,y). Plug into op.
+  auto &srcBlock = region().front();
+  BlockAndValueMapping bvmF; // f(x,y)
+  BlockAndValueMapping bvmR; // f(y,x)
+  {
+    // Save previous insertion point. Continue within loop body.
+    OpBuilder::InsertionGuard guard(b);
+    b.setInsertionPointToEnd(&scfFor.getRegion().front());
+    for (auto it : llvm::zip(srcBlock.getArguments(),
+                             ValueRange{loopCarryValues[0], kValue})) {
+      bvmF.map(std::get<0>(it), std::get<1>(it));
+    }
+    for (auto it : llvm::zip(srcBlock.getArguments(),
+                             ValueRange{kValue, loopCarryValues[0]})) {
+      bvmR.map(std::get<0>(it), std::get<1>(it));
+    }
+    for (auto &blockOp : srcBlock.without_terminator()) {
+      b.clone(blockOp, bvmF);
+      b.clone(blockOp, bvmR);
+    }
+    Value forwardCmpRes = bvmF.lookup(srcBlock.getTerminator()->getOperand(0));
+    Value reverseCmpRes = bvmR.lookup(srcBlock.getTerminator()->getOperand(0));
+
+    // Check value equality using strictly weak ordering from the region:
+    //   f(x,y) --> forwardCmpRes
+    //   f(y,x) --> reverseCmpRes
+    //   if forwardCmpRes == reverseCmpRes then select which came first
+    Value cmpEqValRes = b.create<arith::CmpIOp>(loc, arith::CmpIPredicate::eq,
+                                                forwardCmpRes, reverseCmpRes);
+    Value cmpEqIndRes = b.create<arith::CmpIOp>(loc, arith::CmpIPredicate::slt,
+                                                loopCarryValues[1], kIndex);
+    Value combinedCmpEqRes =
+        b.create<arith::AndIOp>(loc, cmpEqValRes, cmpEqIndRes);
+    // True if N > K or N came before K
```

Comment:
I don't really follow the logic... IIUC, `forwardCmpRes == reverseCmpRes` is true only if this is the first K iterations? If this is correct, let's rename `cmpEqValRes` to a meaningful name. That would helps readability and understand the logic better. If this is the case, can we compare ivs?

The other question, why do we compare the indices? My understanding is

1. If this is the first K iterations, let's select higher/min value.
2. If not, let's compare and select non higher/min value.

Reply:
Thanks, I now know the logic better!